File: C/Spinnaker/Neuronscape/LibNeuronscape/DB_ClientObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Neuronscape
{
    public class DB_ClientObject
    {
        private object locker = new object();
        private Dictionary<UInt32, Record_ClientObject> DB_ID = new Dictionary<UInt32, Record_ClientObject>();
        private bool DirtyBit = false;

        public DB_ClientObject()
        {
            this.Clear();
        }
        public void Clear()
        {
            lock (locker)
            {
                this.DB_ID.Clear();
                this.DirtyBit = false;
            }
        }
        public void SetDirtyBit()
        {
            lock (locker)
            {
                this.DirtyBit = true;
            }
        }
        public void ClearDirtyBit()
        {
            lock (locker)
            {
                this.DirtyBit = false;
            }
        }
        public bool IsDirty()
        {
            lock (locker)
            {
                return this.DirtyBit;
            }
        }
        public UInt32 NumberObjects
        {
            get
            {
                UInt32 NumObjects;

                lock (locker)
                {
                    NumObjects = (UInt32) DB_ID.Count;
                }

                return NumObjects;
            }
        }
        public List<Record_ClientObject> FetchAll()
        {
            List<Record_ClientObject> Objects = new List<Record_ClientObject>();

            lock (locker)
            {
                foreach (KeyValuePair<UInt32, Record_ClientObject> entry in DB_ID)
                {
                    Objects.Add(entry.Value);
                }
            }

            return Objects;
        }
        public bool FindByID(UInt32 ID, out Record_ClientObject Object)
        {
            lock (locker)
            {

                if (this.DB_ID.ContainsKey(ID))
                {
                    Object = this.DB_ID[ID];
                    return true;
                }
                else
                {
                    Object = null;
                    return false;
                }
            }

        }

        public bool AddUpdateRecord(UInt32 ID, CartesianVector Position, double Theta, double Phi, byte Red, byte Green, byte Blue, byte Alpha, double Energy, UInt16 Flags)
        {
            Record_ClientObject Record = new Record_ClientObject();

            Record.ID = ID;
            Record.Position = Position;
            Record.Theta = Theta;
            Record.Phi = Phi;
            Record.Red = Red;
            Record.Green = Green;
            Record.Blue = Blue;
            Record.Brightness = Alpha;
            Record.Energy = Energy;
            Record.FLAGS = Flags;

            return this.AddUpdateRecord(ID, Record);
        }
        public bool AddUpdateRecord(UInt32 ID, Record_ClientObject Record)
        {
            bool ReturnValue = false;

            lock (locker)
            {
                if (this.DB_ID.ContainsKey(ID))
                {
                    this.DB_ID[ID] = Record;
                }
                else
                {
                    this.DB_ID.Add(Record.ID, Record);
                }

                this.DirtyBit = true;
            }            

            return ReturnValue;
        }

        // Delete Record Functions
        public bool DeleteRecordByID(UInt32 ID)
        {
            lock (locker)
            {
                if (!this.DB_ID.ContainsKey(ID))
                {
                    return false;
                }

                 this.DB_ID.Remove(ID);

                 this.DirtyBit = true;
            }
            return true;
        }
    }
}

File: C/Neuronscape/Common/DB_ClientObject.cpp
#include "DB_ClientObject.h"

DB_ClientObject::DB_ClientObject()
{
	this->m_DB_Mutex = new wxMutex(wxMUTEX_RECURSIVE);
	this->Clear();
	this->m_DirtyBit = false;
}
DB_ClientObject::~DB_ClientObject()
{
	this->Clear();
	delete this->m_DB_Mutex;
}
void DB_ClientObject::Clear()
{
	wxMutexLocker Lock(*this->m_DB_Mutex);

	if (this->m_ObjectDB_KEYID.size() != 0)
	{
		DBRecord_ClientObject *Record;
		// Loop through and remove all attached objects, free the memory
		for (DBClientObject_KEYUINT::iterator iter = this->m_ObjectDB_KEYID.begin(); iter != this->m_ObjectDB_KEYID.end(); ++iter)
		{
			Record = iter->second;
			delete Record;
		}
		// Clear the entire list
		this->m_ObjectDB_KEYID.clear();
	}
}
void DB_ClientObject::ClearDirtyBit()
{
	this->m_DirtyBit = false;
}
bool DB_ClientObject::IsDirty()
{
	return this->m_DirtyBit;
}
uint32_t DB_ClientObject::GetNumberOfObjects()
{
	return this->m_ObjectDB_KEYID.size();
}
void DB_ClientObject::FetchAll(std::vector<DBRecord_ClientObject> &Records)
{
	wxMutexLocker Lock(*this->m_DB_Mutex);

	if (this->m_ObjectDB_KEYID.size() != 0)
	{
		for (DBClientObject_KEYUINT::iterator iter = this->m_ObjectDB_KEYID.begin(); iter != this->m_ObjectDB_KEYID.end(); ++iter)
		{
			Records.push_back(*iter->second);
		}
	}
	return;
}
bool DB_ClientObject::FindByID(uint32_t ID, DBRecord_ClientObject &Record)
{
	wxMutexLocker Lock(*this->m_DB_Mutex);

	if (ID == 0)
	{
		return false;
	}

	DBClientObject_KEYUINT::iterator iter = this->m_ObjectDB_KEYID.find(ID);

	if (iter != this->m_ObjectDB_KEYID.end()) // If this is the case the object was found in the list
	{
		Record = *iter->second;
		return true;
	}
	else
	{
		return false;
	}
}
bool DB_ClientObject::AddObject(uint32_t ID, CartesianVector Position, double theta, double phi, unsigned char Red, unsigned char Green, unsigned char Blue, unsigned char Brightness, double Energy, uint16_t Flags)
{
	wxMutexLocker Lock(*this->m_DB_Mutex);

	DBRecord_ClientObject Record;

	if (this->FindByID(ID, Record))
	{
		return false;
	}

	DBRecord_ClientObject *NewRecord = new DBRecord_ClientObject(ID,Position,theta,phi,Red,Green,Blue,Brightness,Energy,Flags);

	std::pair<DBClientObject_KEYUINT::iterator,bool> ret = this->m_ObjectDB_KEYID.insert(std::pair<uint32_t, DBRecord_ClientObject*>(ID,NewRecord));

	if (ret.second == false)
	{
		delete NewRecord;
		return false;
	}

	this->m_DirtyBit = true;
	return true;
}
bool DB_ClientObject::UpdateObject(uint32_t ID, CartesianVector Position, double theta, double phi, unsigned char Red, unsigned char Green, unsigned char Blue, unsigned char Brightness, double Energy, uint16_t Flags)
{
	wxMutexLocker Lock(*this->m_DB_Mutex);

	DBRecord_ClientObject *Record = NULL;

	DBClientObject_KEYUINT::iterator iter = this->m_ObjectDB_KEYID.find(ID);

	if (iter == this->m_ObjectDB_KEYID.end()) // If this is the case the object was found in the list
	{
		return false;
	}

	Record = iter->second;

	Record->ID = ID;
	Record->Position = Position;
	Record->Theta = theta;
	Record->Phi = phi;
	Record->Red = Red;
	Record->Green = Green;
	Record->Blue = Blue;
	Record->Brightness = Brightness;
	Record->Energy = Energy;
	Record->Flags = Flags;

	this->m_DirtyBit = true;
	return true;
}
bool DB_ClientObject::AddUpdateObject(uint32_t ID, CartesianVector Position, double theta, double phi, unsigned char Red, unsigned char Green, unsigned char Blue, unsigned char Brightness, double Energy, uint16_t Flags)
{
	wxMutexLocker Lock(*this->m_DB_Mutex);

	DBRecord_ClientObject Record;

	if (this->FindByID(ID,Record))
	{
		this->UpdateObject(ID,Position,theta,phi,Red,Green,Blue,Brightness, Energy, Flags);
	}
	else
	{
		this->AddObject(ID,Position,theta,phi,Red,Green,Blue,Brightness, Energy, Flags);
	}

	return true;
}
bool DB_ClientObject::DeleteObject(uint32_t ID)
{
	wxMutexLocker Lock(*this->m_DB_Mutex);

	DBRecord_ClientObject *Record = NULL;

	DBClientObject_KEYUINT::iterator iter = this->m_ObjectDB_KEYID.find(ID);

	if (iter == this->m_ObjectDB_KEYID.end()) // If this is the case the object was not found in the list
	{
		return false;
	}

	Record = iter->second;

	this->m_ObjectDB_KEYID.erase(ID);

	delete Record;

	this->m_DirtyBit = true;

	return true;
}
